master/internal/stream/test_util.go

```diff
@@ -0,0 +1,254 @@
+	return i
+}
+
+// mockSocket implements WebsocketLike and stores all messages received from streaming.
+type mockSocket struct {
+	inbound  chan interface{}
+	outbound chan *StartupMsg
+	closed   chan struct{}
+}
+
+// newMockSocket creates a new instance mockSocket and initialize it's conditional variables.
+func newMockSocket() *mockSocket {
+	return &mockSocket{
+		inbound:  make(chan interface{}, channelBufferSize),
+		outbound: make(chan *StartupMsg, channelBufferSize),
+		closed:   make(chan struct{}),
+	}
+}
+
+// WriteOut synchrounously appends the StartupMsg to outbound messages.
+func (s *mockSocket) WriteOutbound(startup *StartupMsg) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case s.outbound <- startup:
+		return nil
+	}
+}
+
+// ReadJSON implements WebsocketLike's ReadJSON(), blocks until able to read outbound messages off the mockSocket
+// or the mockSocket is closed.
+func (s *mockSocket) ReadJSON(data interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case msg := <-s.outbound:
+		targetMsg, ok := data.(*StartupMsg)
+		if !ok {
+			return fmt.Errorf("target message type is not a pointer to StartupMsg")
+		}
+		*targetMsg = *msg
+		return nil
+	}
+}
+
+// Write implements WebsocketLike's Write(), appends the data to mockSocket's inbound messages.
+func (s *mockSocket) Write(data interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case s.inbound <- data:
+		return nil
+	}
+}
+
+// ReadData synchrounously reads an inbound message off the mockSocket.
+func (s *mockSocket) ReadIncoming(data *interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case msg := <-s.inbound:
+		*data = msg
+		return nil
+	}
+}
+
+// ReadUntil reads until the terminationMsg has been read.
+func (s *mockSocket) ReadUntil(
+	t *testing.T,
+	testCaseDescription string, // XXX (corban): we should use subtests instead, this won't be necessary
+	data *[]interface{},
+	terminationMsg interface{},
+) {
+	var msg interface{}
+ReadLoop:
+	for {
+		if reflect.TypeOf(msg) == reflect.TypeOf(terminationMsg) {
+			switch typedMsg := msg.(type) {
+			case stream.UpsertMsg:
+				if err := validateUpsertMsg(typedMsg.Msg, terminationMsg.(stream.UpsertMsg).Msg); err == nil {
+					break ReadLoop
+				}
+			case string:
+				if typedMsg == terminationMsg.(string) {
+					break ReadLoop
+				}
+			case SyncMsg:
+				if typedMsg.SyncID == terminationMsg.(SyncMsg).SyncID {
+					break ReadLoop
+				}
+			}
+		}
+		if err := s.ReadIncoming(&msg); err != nil {
+			t.Errorf("%s: %s", testCaseDescription, err)
+		}
+		*data = append(*data, msg)
+	}
+}
+
+// Close closes the mockSocket.
+func (s *mockSocket) Close() {
+	close(s.closed)
+}
+
+func splitMsgs[M stream.Msg](
```

Comment:
was the change to strings going to land this PR or next?

Reply:
The next PR :D